### PR TITLE
fix(router): fix incorrectly calculated priorities in traditional_compat

### PR DESCRIPTION
**Pull request by hanshuebner**
### Summary

When prefix and regex routes were mixed or prefixes of different lengths were specified in one route, traditional_compat assigned the highest priority of all paths to the route for all paths, causing unexpected routing matches that were incompatible with the traditional router.

With this fix, each multi-path route is split into multiple routes with separate priorities.

### Checklist

- [X] The Pull Request has tests
- [x] There's an entry in the CHANGELOG
- [x] There is a user-facing [docs PR](https://github.com/Kong/docs.konghq.com/pull/5402)

### Issue reference

KAG-727
[KAG-1446](https://konghq.atlassian.net/browse/KAG-1446)

[KAG-1446]: https://konghq.atlassian.net/browse/KAG-1446?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: kong/router/compat.lua [hunk 1]
```diff
@@ -249,27 +252,34 @@ local function get_priority(route)
...
   local regex_url = false
 
   if not is_empty_field(paths) then
     match_weight = match_weight + 1
 
-    for _, p in ipairs(paths) do
-      if is_regex_magic(p) then
-        regex_url = true
-
+    for index, p in ipairs(paths) do
```
Comment (samugi) [inline kong/router/compat.lua:261]: if we are enforcing one path per route, can we just assert that `#route.paths == 1` here and avoid this loop? 

Or somewhere in atc.lua this gets called with a route that can have many paths?

Comment (hanshuebner) [reply kong/router/compat.lua:261]: You are right.  I kind of expect that someone points out the possibility to group regular expressions and paths of equal lengths, so I left the loop in place in preparation of that.

File: CHANGELOG.md [hunk 2]
```diff
@@ -105,6 +105,9 @@
   Thanks [@PidgeyBE](https://github.com/PidgeyBE) for contributing this change.
   [#10595](https://github.com/Kong/kong/pull/10595)
   [#10204](https://github.com/Kong/kong/pull/10204)
+- Make `traditional_compat` mode more compatible to `traditional` by splitting routes with
+  multiple paths into multiple atc routes with separate priorities.
```
Comment (hbagdi) [inline CHANGELOG.md:109]: Please elaborate on the impact on the end-user here.
The user should be able to read a single line and understand the change of behavior.
Also, please note this in the "breaking change/behavior" section since a user could already be relying on such behavior and this change could surprise them.

File: spec/01-unit/08-router_spec.lua [hunk 3]
```diff
@@ -2181,13 +2181,22 @@ for _, flavor in ipairs({ "traditional", "traditional_compatible", "expressions"
 
         lazy_setup(function()
           use_case = {
+            -- plain
+            {
+              service = service,
+              route   = {
+                id = "e8fb37f1-102d-461e-9c51-6608a6bb8100",
+                paths = {
+                  "/plain/a.b.c", -- /plain/a.b.c
```
[the remark this answers is not in the document]
Comment (bungle) [reply spec/01-unit/08-router_spec.lua:2190]: I checked this. It looks like we don't support anymore non-normalized forms to be entered in db. So all uris there should be in normalized form I think.

[the remark this answers is not in the document]
Comment (hanshuebner) [reply spec/01-unit/08-router_spec.lua:2190]: It seems that it will be best to remove the percent encoded path from the test to avoid this to be confusing in the future.

[the remark this answers is not in the document]
Comment (hanshuebner) [reply spec/01-unit/08-router_spec.lua:2190]: Actually, the test was valuable but wrong.  It wanted to verify that configured prefix paths are no longer percent-decoded, but it did not actually verify that.  I've fixed it.

File: kong/router/compat.lua [hunk 4]
```diff
@@ -74,7 +74,7 @@ local function get_expression(route)
     tb_insert(out, gen)
   end
 
-  local gen = gen_for_field("tls.sni", OP_EQUAL, snis, function(op, p)
+  gen = gen_for_field("tls.sni", OP_EQUAL, snis, function(_, p)
```
Comment (chronolaw) [inline kong/router/compat.lua:77]: It is not a related change, could we do it in another PR?

File: kong/router/compat.lua [hunk 5]
```diff
@@ -128,7 +128,7 @@ local function get_expression(route)
     end)
   end
 
-  local gen = gen_for_field("http.path", function(path)
+  gen = gen_for_field("http.path", function(path)
```
Comment (chronolaw) [inline kong/router/compat.lua:131]: It is not a related change, could we do it in another PR?

File: kong/router/compat.lua [hunk 6]
```diff
@@ -249,27 +249,34 @@ local function get_priority(route)
...
-      if is_regex_magic(p) then
-        regex_url = true
-
+    for index, p in ipairs(paths) do
+      if index == 1 then
+        if is_regex_magic(p) then
+          regex_url = true
+        else
+          uri_length = #p
+        end
```
Comment (chronolaw) [inline kong/router/compat.lua:264]: add a blank line before `else`

File: kong/router/compat.lua [hunk 7]
```diff
@@ -249,27 +249,34 @@ local function get_priority(route)
...
+          assert(#p == uri_length, "cannot mix different length prefixes in get_priority")
+        end
       end
     end
   end
 
-  local match_weight   = lshift_uint64(match_weight, 61)
-  local headers_count  = lshift_uint64(headers_count, 52)
+  match_weight   = lshift_uint64(match_weight, 61)
+  headers_count  = lshift_uint64(headers_count, 52)
```
Comment (chronolaw) [inline kong/router/compat.lua:276]: It is not a related change, could we do it in another PR?

File: kong/router/atc.lua [hunk 8]
```diff
@@ -443,6 +442,9 @@ function _M:select(req_method, req_uri, req_host, req_scheme,
 
   local service = self.services[uuid]
   local matched_route = self.routes[uuid]
+  if matched_route.original_route then
```
Comment (chronolaw) [inline kong/router/atc.lua:445]: could be:
```
local matched_route = matched_route.original_route or self.routes[uuid]
```

File: kong/router/compat.lua [hunk 9]
```diff
@@ -295,8 +302,58 @@ local function get_exp_and_priority(route)
...
-function _M.new(routes, cache, cache_neg, old_router)
-  return atc.new(routes, cache, cache_neg, old_router, get_exp_and_priority)
+-- group array-like table t by the function f, returning a table mapping from
+-- the result of invoking f on one of the elements to the actual elements.
+local function group_by(t, f)
+  local result = {}
+  for _, value in ipairs(t) do
+    local key = f(value)
+    if result[key] then
+      result[key][#result[key] + 1] = value
```
Comment (chronolaw) [inline kong/router/compat.lua:312]: `table.insert` seems to have an equal result.

File: kong/router/compat.lua [hunk 10]
```diff
@@ -295,8 +302,58 @@ local function get_exp_and_priority(route)
...
+
+  assert(rs.route.paths)
+  local grouped_paths = group_by(
+    rs.route.paths,
+    function(path)
+      return is_regex_magic(path) or #path
+    end
+  )
+  for _, paths in pairs(grouped_paths) do
+    local cloned_route = tablex.deepcopy(rs)
```
Comment (flrgh) [inline kong/router/compat.lua:336]: nit/non-blocker: `kong.tools.utils` has a `deep_copy()` function. Could be wise to use it instead for consistency.

File: kong/router/compat.lua [hunk 11]
```diff
@@ -295,8 +302,58 @@ local function get_exp_and_priority(route)
...
+    split_rs[#split_rs + 1] = rs
+    return
+  end
+
+  assert(rs.route.paths)
+  local grouped_paths = group_by(
+    rs.route.paths,
+    function(path)
+      return is_regex_magic(path) or #path
+    end
```
Comment (flrgh) [inline kong/router/compat.lua:333]: I don't know if luajit will be clever enough to optimize away the closure creation here, but unless I'm mistaken this can just as easily be turned into a regular `local function` at the module scope rather than a closure.

File: kong/router/compat.lua [hunk 12]
```diff
@@ -295,8 +302,58 @@ local function get_exp_and_priority(route)
...
+
+-- split routes into multiple routes, one for each prefix length and one for all
+-- regular expressions
+local function split_route_by_path_into(rs, split_rs)
+  if is_empty_field(rs.route.paths) or #rs.route.paths == 1 then
+    split_rs[#split_rs + 1] = rs
+    return
+  end
+
+  assert(rs.route.paths)
```
Comment (flrgh) [inline kong/router/compat.lua:328]: This looks to be guarded against by `is_empty_field(rs.route.paths)` in the clause on line 323. If that weren't the case then the second half of the conditional (`or #rs.route.paths == 1`) would be "dangerous" due to not performing a nil check before attempting to read the table length.

File: kong/router/compat.lua (hunk 10 above)
Comment (flrgh) [inline kong/router/compat.lua:336]: I think the biggest concern I have about the PR is right here, since we know there are users out there with service/route counts in the mid 5 figures. I think it's a case where it's worth splitting some hairs to see what we can do to cut down on copying tables.

For one, I wonder if it's possible to get away with just copying the route and re-using the service?

```lua
local cloned_route = {
  route = tablex.deepcopy(rs.route),
  original_route = rs.route,
  service = rs.service,
}
cloned_route.route.paths = paths
cloned_route.route.id = utils.uuid()
```

Additionally, as long as we're reasonably certain that `atc` doesn't modify the input, I think you can likely use a [shallow clone](https://github.com/Kong/kong/blob/c0a7eb816dcdf02b5ea63767176d7420df1059ff/kong/tools/utils.lua#L612-L624) of the route rather than a deep copy. The route entity has a bunch of table fields, so a shallow clone will be a lot cheaper than a deep copy:

```lua
local cloned_route = {
  route = utils.shallow_copy(rs.route),
  original_route = rs.route,
  service = rs.service,
}
cloned_route.route.paths = paths
cloned_route.route.id = utils.uuid()
```

There is _some_ maintenance risk involved with this (what if a new property is added to this `rs` table?), but thorough test coverage should/would catch that.

**Review by flrgh — COMMENTED** [review]
Nicely done! I left a few small suggestions, and I'd like to know your thoughts about the [table copy comment that I raised](https://github.com/Kong/kong/pull/10615/files#r1179496259).

File: kong/router/compat.lua (hunk 11 above)
Comment (hanshuebner) [reply kong/router/compat.lua:333]: There is no difference between the two in my measurements, so I'm leaving it as-is.

File: kong/router/compat.lua (hunk 10 above)
Comment (hanshuebner) [reply kong/router/compat.lua:336]: Thank you for your review and the optimization suggestion.

I've opted to check the `rs` table for unexpected content while processing it.  According to my reading, `atc` does not update the routes itself, so sharing should be fine and I've gone with the shallow copy.

File: kong/router/compat.lua [hunk 13]
```diff
@@ -295,8 +305,62 @@ local function get_exp_and_priority(route)
...
+    end
+  )
+  for _, paths in pairs(grouped_paths) do
+    local cloned_route = {
+      route = utils.shallow_copy(rs.route),
+      service = rs.service,
+    }
+    cloned_route.route.original_route = rs.route
+    cloned_route.route.paths = paths
+    cloned_route.route.id = utils.uuid()
```
Comment (bungle) [inline kong/router/compat.lua:346]: I am wondering what this would do to incremental rebuilds?

**Review by bungle — COMMENTED** [review]
Some early comments.

File: kong/router/compat.lua (hunk 13 above)
Comment (hanshuebner) [reply kong/router/compat.lua]: I've changed the id generation for the split routes so that a uuidv5 is used, based on the original route ID and the index of the path.  That way, incremental rebuilds will not discard and rebuild these split routes if they were not changed.

**Comment by hanshuebner** [comment]
@bungle can I get a final review and approval, please?

**Review by bungle — APPROVED** [review]
LGTM
